finance/server/src/main/java/com/google/finapp/SpannerDaoPGImpl.java

```diff
@@ -0,0 +1,367 @@
+    this.databaseClient = databaseClient;
+  }
+
+  public void createCustomer(ByteArray customerId, String name, String address)
+      throws StatusException {
+    try {
+      this.databaseClient
+          .readWriteTransaction()
+          .run(
+              transaction -> {
+                Statement statement =
+                    Statement.newBuilder(
+                            "INSERT INTO Customer\n"
+                                + "(CustomerId, Name, Address)\n"
+                                + "VALUES\n"
+                                + "(@customerId, @name, @address)")
+                        .bind("customerId")
+                        .to(customerId)
+                        .bind("name")
+                        .to(name)
+                        .bind("address")
+                        .to(address)
+                        .build();
+                transaction.executeUpdate(statement);
+                return null;
+              });
+    } catch (SpannerException e) {
+      throw Status.fromThrowable(e).asException();
+    }
+  }
+
+  public void createAccount(ByteArray accountId, AccountStatus accountStatus, BigDecimal balance)
+      throws StatusException {
+    try {
+      this.databaseClient
+          .readWriteTransaction()
+          .run(
+              transaction -> {
+                Statement statement =
+                    Statement.newBuilder(
+                            "INSERT INTO Account\n"
+                                + "(AccountId, AccountStatus, Balance, CreationTimestamp)\n"
+                                + "VALUES\n"
+                                + "($1, $2, $3, SPANNER.PENDING_COMMIT_TIMESTAMP())")
+                        .bind("p1")
+                        .to(accountId)
+                        .bind("p2")
+                        .to(accountStatus.getNumber())
+                        .bind("p3")
+                        .to(
+                            Value.pgNumeric(
+                                balance.toString())) // Numeric requires special handling
+                        .build();
+                transaction.executeUpdate(statement);
+
+                return null;
+              });
+    } catch (SpannerException e) {
+      System.out.println(e.getMessage());
+      throw Status.fromThrowable(e).asException();
+    }
+  }
+
+  public void createCustomerRole(
+      ByteArray customerId, ByteArray accountId, ByteArray roleId, String roleName)
+      throws StatusException {
+
+    try {
+      this.databaseClient
+          .readWriteTransaction()
+          .run(
+              transaction -> {
+                Statement statement =
+                    Statement.newBuilder(
+                            "INSERT INTO CustomerRole\n"
+                                + "(CustomerId, AccountId, RoleId, Role)\n"
+                                + "VALUES\n"
+                                + "($1, $2, $3, $4)")
```

Comment:
Related to other comment from Panos. If named parameters work, I'd recommend using them everywhere.

Reply:
Unfortunately, they don't :(